Riot/Modules/Authentication/AuthenticationCoordinator.swift

```diff
@@ -81,11 +90,157 @@ final class AuthenticationCoordinator: NSObject, AuthenticationCoordinatorProtoc
     func continueSSOLogin(withToken loginToken: String, transactionID: String) -> Bool {
         authenticationViewController.continueSSOLogin(withToken: loginToken, txnId: transactionID)
     }
+    
+    // MARK: - Private
+    
+    private func showLoadingAnimation() {
+        let loadingViewController = LaunchLoadingViewController()
+        loadingViewController.modalPresentationStyle = .fullScreen
+        
+        // Replace the navigation stack with the loading animation
+        // as there is nothing to navigate back to.
+        navigationRouter.setRootModule(loadingViewController)
+    }
+    
+    private func presentCompleteSecurity(with session: MXSession) {
+        let isNewSignIn = true
+        let keyVerificationCoordinator = KeyVerificationCoordinator(session: session, flow: .completeSecurity(isNewSignIn))
+        
+        keyVerificationCoordinator.delegate = self
+        let presentable = keyVerificationCoordinator.toPresentable()
+        presentable.presentationController?.delegate = self
+        navigationRouter.present(presentable, animated: true)
+        keyVerificationCoordinator.start()
+        add(childCoordinator: keyVerificationCoordinator)
+    }
+    
+    private func authenticationDidComplete() {
+        completion?(.didComplete(authenticationViewController.authType))
```

Comment:
I see this is called from a lot of places, even on failures. Is that expected?

Reply:
Yes in as much as this is how `AuthenticationViewController` previously worked (calling `dismiss`).

The calls from `sessionStateDidChange` are all once the session is running and the failures are only crypto/cross-signing related so should all be handled from the home tab once the onboarding has completed. There is probably scope for improvement once the auth flow is re-written though.